process_hierarchy.py

```diff
@@ -0,0 +1,478 @@
+            )
+            group_file.write(f'"{l4_rec["name"]}", "{unique_key}" \n')
+
+    # constrcut pi2project
+    with open("pi2project.csv", "w", encoding="utf-8") as pi2project_file:
+        for l5 in hierarchy["cloud-project"].values():
+            l4_rec = hierarchy["cloud-project"][l5["parent_id"]]
+            l4_id = l4_rec["id"]
+            pi2project_file.write(
+                f"{hierarchy['cloud-project'][l4_id]['name']}, {l5['name']}\n"
+            )
+
+    # construct the names (rename projects)
+
+
+def find_hierarchy_id(name, dictionary):
+    """returns the hierarchy_id if found in the dictionary"""
+    for rec in dictionary.values():
+        if name == rec["name"]:
+            return rec["id"]
+    return None
+
+
+def get_hierarchy_from_db(cursor):
+    """This fetches the data from the database and only keeps the most recent record based on its id"""
+    hierarchy = {}
+    hierarchy["institution"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='institution' order by hr.create_ts",
+        None,
+    )
+    hierarchy["field-of-science"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='field-of-science' order by hr.create_ts",
+        None,
+    )
+    hierarchy["pi"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='pi' order by hr.create_ts",
+        None,
+    )
+    hierarchy["project"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='openstack-project' or hr.type='openshift-project' order by hr.create_ts",
+        None,
+    )
+    hierarchy["cloud-project"] = create_hierarchy_dictionary(
+        cursor,
+        "select * from hierarchy_db.hierarchy_rec hr where hr.type='openstack-project' order by hr.create_ts",
+        None,
+    )
+    return hierarchy
+
+
+def process_expected_data(cursor, hierarchy):
+    """This fits idealized data into the hierarchy from the database
+
+    This is currently needed as we are not getting the data from
+    keycloak yet.
+    """
+    data = get_data()
+    for record in data:
+        match record["type"]:
+            case "institution":
+                process_record(cursor, record, hierarchy["institution"])
+            case "field-of-science":
+                process_record(cursor, record, hierarchy["field-of-science"])
+            case "pi":
+                process_record(cursor, record, hierarchy["pi"])
+            case "openshift-project":
+                process_record(cursor, record, hierarchy["project"])
+            case "openstack-project":
+                process_record(cursor, record, hierarchy["project"])
+                record["parent_id"] = record["id"]
+                process_record(cursor, record, hierarchy["cloud-project"])
+    return 1
```

Comment:
Is there any reason this function is returning anything at all?

Reply:
eliminated